### ISSUE TITLE
ClassCastException in iOS11

### ISSUE THREAD

**Issue by HoneyLuka** [google/j2objc#889]
Hi,
  I will use fake code to describe the problem.

I use j2objc in a framework【FrameworkA】.
a protocol【ProcotolA】(translate from java interface)
and I have a class like this
```
class ClassA : NSObject, ProtocolA {
    //ProtocolA implemention code.
}
```

in java code have some code like this
```
//obj is ClassA instance
public void method(Object obj) {
    otherMethod((ProtocolA) obj);
}
```

when I use frameworkA in my project it says
```
java.lang.ClassCastException: Cannot cast object of type FrameworkA.ClassA to ProtocolA
```

My j2objc version is 2.0.4. Swift version is 3.2. (use 3.1 before iOS11)
But before iOS11 update it is ok, very weird. Can you give me some suggestions about this? 
Thank you very much.

**Comment by clementmangin** [comment]
I think I'm faced with a similar issue, which is related to Xcode 9 (the error occurs in both iOS 10 and iOS 11, and did not occur with Xcode 8).

I have a Java class A which extends an abstract class B, and the class B implements an interface C.
I get a ClassCastException when trying to cast A to C:

`id<C> obj2 = (id<C>) cast_check(obj1, C_class_());`

In `cast_check` (in J2Objc_source.h), I've found that [cls isInstance:p] returns nil.
cls being an interface, I've looked at the implementation of `isInstance:` and `ConformsToProtocol` in IOSProtocolClass.m, and put a breakpoint on the `[cls isInstance:p]` line to check what was happening. Here is what I got:

What I've found in the debug console of Xcode:

```
(lldb) po [p java_getClass]
class A

(lldb) po [[p java_getClass] getSuperclass]
class B

(lldb) po [[[p java_getClass] getSuperclass] getInterfacesInternal]
[]
```

The bug seems to be in `getInterfacesInternal`, which should return a list containing my interface C.

I hope this helps understand the issue!

**Comment by HoneyLuka** [comment]
@clementmangin Thanks for your reply. 
Maybe it is related with xcode version. 
Now I use a trick to resolve this problem. add this code below to java file
```
/*-[
#define cast_check(x, y) my_cast_check(x, y)
__attribute__((always_inline)) inline id my_cast_check(id __unsafe_unretained p, IOSClass *cls) {
  return p;
}
]-*/
```
this will replace cast_check method implemention.
Hope can help.

**Comment by tomball** [comment]
Defining J2OBJC_DISABLE_CAST_CHECKS will also disable cast checking. To do
so from Xcode, in the target's Build Settings, click on the + button at the
top (to the left of the Find widget, under the Build Phases label), and set
any value (it's only the definition itself that's checked).

**Comment by kstanger** [comment]
This looks similar to the issue discussed here: https://groups.google.com/d/topic/j2objc-discuss/avy79pz1XxU/discussion

**Comment by dmoss18** [comment]
I added the `J2OBJC_DISABLE_CAST_CHECKS` build setting but it doesn't seem to work. I added it to my app's target and set the value to 1. It's a User-Defined setting right?

However, the workaround mentioned here https://groups.google.com/d/topic/j2objc-discuss/avy79pz1XxU/discussion did work (adding a `@protocol(MyInterface)`)

**Comment by tomball** [comment]
Where did you add the @protocol(MyInterface) workaround? I'm trying to
figure out if the translator can create those, but am not sure where the
best place for them is.

**Comment by dmoss18** [comment]
I created a .m file with the following content:

import “path/to/myprotocol.h”
void workaround() {
@protocol(MyProtocol);
}

And added that file to the projects compiled sources.

**Comment by dfsilva** [comment]
I am having this problem too.

**Comment by tomball** [comment]
Did you try the workaround other developers have posted?

**Comment by dfsilva** [comment]
Yes, I solved with solution posted by @HoneyLuka . tks

**Comment by dfsilva** [comment]
Somebody know if this properties is working to solve this problem? Because I am having the cast problem in another points of my app and I would like to fix the problem in one single point. I have tried to define this properties J2OBJC_DISABLE_ALL_CHECKS and J2OBJC_DISABLE_ARRAY_CHECKS on my Build Settings -> User-Defined but didn't work. Somebody can help me? I am using the version 2.0.5 of j2objc.

**Comment by danielCarlosCE** [comment]
is **j2objc** planning to solve this problem?

**Comment by tomball** [comment]
It's a Swift importer problem -- if anyone can describe what code j2objc
can generate, we'd love to hear about it. Generating a .m file with an
@protocol declaration for all Java interfaces will bloat all apps that use
j2objc. Generating a single .m if the Swift importer bug bites your project
has a much smaller impact.

**Comment by danielCarlosCE** [comment]
Was it really confirmed as a Swift problem?

**Comment by tomball** [comment]
The j2objc translator generates Objective C .h files which are read by the
Swift importer. These files worked before iOS 11, and broke when it was
released.

While we're happy to change what's generated if iOS 11's version of Swift
needs something different, we've found no documentation describing what
that change should be. If anyone knows of new attributes or annotations
that a protocol declaration should have to be Swift-compatible, please
share them.

**Comment by flovilmart** [comment]
I can confirm here that the protocol loading workaround is doing it's magic, not sure what's going on, and will keep on investigate

**Comment by flovilmart** [comment]
@tomball I reached out to the swift compiler team (https://twitter.com/uint_min/status/971105499415392257) and for them it smells like a bug. I'll try to open an issue and let you guys know.

**Comment by tomball** [comment]
Thanks for following up on this!

**Comment by flovilmart** [comment]
For the sake of documentation, here's the demo project https://github.com/flovilmart/TestingObjCProtocolIssue

**Comment by mayur8657** [comment]
@flovilmart The protocol workaround is not helping me.

I have an interface "InterfaceTypeA" and a class that implements this interface "ClassTypeA".

I have a method that accepts parameter id<InterfaceTypeA> and when I pass it an object of "ClassTypeA". I get a class cast exception.

Fatal Exception: JavaLangClassCastException
java.lang.ClassCastException: Cannot cast object of type ClassTypeA to InterfaceTypeA.

I added a MyProtocolWorkaround.m file and added it to my Compile Sources of the project.

But it doesn't change anything.

What's wonderful? The exception occurs only with Release build configuration and not in Debug build configuration of my Scheme. Anyone any idea whats going wrong? Why the workaround fails for me? And why my debug build works even without the workaround?

**Comment by flovilmart** [comment]
@mayur8657 I haven’t personally tested with release configurations and optimizations any chance you can share a simple project, or open a PR on the demo project I published with your failing use case?

**Comment by saruye** [comment]
does anybody when this fix will go into a release toolchain? The development toolchains seem to solve the problem for us. But I cannot upload the builds now. I tried to figure out which in which builds this fix is included, but couldn't really figure it out. Does anybody know the process or where to look? 
That is the issue:  https://bugs.swift.org/browse/SR-7130 and that the commit https://github.com/apple/swift/commit/aaa5ce11bb47e129ba654fa65360eb582436232a
But there I only see that it is in the 4.2 branch and in master. Do I need to wait for 4.2 to be released now? Just have to make a decision if we can wait or need to find another workaround.

**Comment by flovilmart** [comment]
There’s a workaround you can use, by generating a dummy file with the protocol imports so all types you need have the 
`@protocol(TypeThatINeed)`.

**Comment by tomball** [comment]
There is no fix waiting to be released, as the bug is in the Swift
importer, which we don't control. So as Florent suggests, use the dummy
file workaround.

If anyone knows of a better workaround, please share it. If anyone knows of
a better Swift compiler than the POS Apple ships, please let us know.

**Comment by saruye** [comment]
I tried the workaround with the dummy file. But somehow that didn't help. We are using the gradle plugin to generate the binary file which is then included into the project with cocoapods. Could that be an issue? I also tried adding a workaround file into our gradle build but also without success. Also the bug only happens when doing release builds. 
First we had the issue in our own code. After refactoring it a bit it stopped. Now we are facing it somewhere deep in guava. 

@tomball i know that this is not an issue with j2objc. I was just asking here if anybody knew where I can find out in which swift build, that can be used for releasing an app, this fix will be released. The development toolchains help us, but they don't get us closer to releasing our app.
